sympy/core/operations.py

```diff
@@ -504,3 +505,117 @@ def args(self):
+
+    >>> @add.register_handler(MatrixExpr)
+    ... def _(*args, **kwargs):
+    ...     return MatAdd(*args, **kwargs)
+
+    >>> @add.register_priority(Expr, MatrixExpr)
+    ... def _(expr, mexpr):
+    ...     return mexpr
+
+    >>> from sympy import MatrixSymbol, Symbol
+    >>> a, A = Symbol('a'), MatrixSymbol('A', 2, 2)
+
+    >>> add(a, A, a) == MatAdd(a, A, a)
+    True
+    """
+
+    def __init__(self, name, fallback, doc=None):
+        self.handler_dispatcher = Dispatcher("%s's handler" % name)
+        super().__init__(name, fallback, doc)
+
+    def register_handler(self, cls):
+        """
+        Register the handler function for a type.
+        """
+        def _(func):
+            self.handler_dispatcher.register(cls)(func)
+            return func
+        return _
+
+    def get_handler(self, cls):
+        """
+        Return the registered handler for the argument of given type.
+        """
+        return self.handler_dispatcher.dispatch(cls)
+
+    def register_priority(self, cls1, cls2):
+        """
+        Register the priority between two types. Registered function must
+        return the argument of higher priority.
+        """
+        def _(func):
+            self.add((cls1, cls2), func)
+            return func
+        return _
+
+    def binary_select(self, cls1, cls2):
+        """
+        Select the class using registered priority. If the priority
+        cannot be determined, return ``None``.
+        """
+        result = self.dispatch(cls1, cls2)
+        if result:
+            return result(cls1, cls2)
+        else:
+            result = self.dispatch(cls2, cls1)
+            if result:
+                return result(cls2, cls1)
+            else:
+                return result
+
+    def find_handler(self, args):
+        """
+        Select the argument with highest priority, and return its handler.
+        """
+        args_data = []
+        for a in args:
+            func, op_priority, handler = type(a), getattr(a, '_op_priority', 0), self.get_handler(type(a))
+            if handler is not None:
+                args_data.append((func, op_priority, handler))
+
+        # No handler detected: use fallback
+        if len(args_data) == 0:
+            return self.fallback
+
+        # Sieve with _op_priority first
+        max_priority = max(args_data, key=lambda x:x[1])[1]
+        sieved_data = [tup for tup in args_data if tup[1]==max_priority]
+
+        # Recursively select with registered binary priority
+        for i, tup in enumerate(sieved_data):
+            if i == 0:
+                result = tup
+            else:
+                selected = self.binary_select(result[0], tup[0])
+                if selected == tup[0]:
+                    result = tup
+                elif selected == result[0]:
+                    pass
+                elif not selected:
+                    # Cannot determine priority: use fallback
+                    return self.fallback
```

Comment:
When would we hit this branch? Maybe it should raise an error...

Reply:
One of this case is where all arguments are non-registered type, such as int (e.g. `add(1, 2)`).  This will be changed by using class attribute as you suggested.

Reply:
I think `add(1, 2)` is handled above

Reply:
Oh, right. I confused. Seems like this shall raise error.